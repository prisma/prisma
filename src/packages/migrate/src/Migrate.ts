import {
  getSchemaPathSync,
  getGenerators,
  uriToCredentials,
  getConfig,
} from '@prisma/sdk'
import chalk from 'chalk'
import Debug from '@prisma/debug'
import fs from 'fs'
import logUpdate from 'log-update'
import path from 'path'
import { MigrateEngine } from './MigrateEngine'
import { EngineResults, EngineArgs } from './types'
import { formatms } from './utils/formatms'
import {
  getDbinfoFromCredentials,
  getDbLocation,
} from './utils/ensureDatabaseExists'
import { enginesVersion } from '@prisma/engines-version'
import { NoSchemaFoundError } from './utils/errors'

const debug = Debug('migrate')
const packageJson = eval(`require('../package.json')`) // tslint:disable-line

export class Migrate {
  get devMigrationsDir(): string {
    return path.join(path.dirname(this.schemaPath), 'migrations/dev')
  }
  public engine: MigrateEngine
  private schemaPath: string
  public migrationsDirectoryPath: string
  constructor(schemaPath?: string, enabledPreviewFeatures?: string[]) {
    this.schemaPath = this.getSchemaPath(schemaPath)
    this.migrationsDirectoryPath = path.join(
      path.dirname(this.schemaPath),
      'migrations',
    )
    this.engine = new MigrateEngine({
      projectDir: path.dirname(this.schemaPath),
      schemaPath: this.schemaPath,
      enabledPreviewFeatures,
    })
  }

  public stop(): void {
    this.engine.stop()
  }

  public getSchemaPath(schemaPathFromOptions?): string {
    const schemaPath = getSchemaPathSync(schemaPathFromOptions)

    if (!schemaPath) {
      throw new NoSchemaFoundError()
    }

    return schemaPath
  }

  public getDatamodel(): string {
    return fs.readFileSync(this.schemaPath, 'utf-8')
  }

  public reset(): Promise<void> {
    return this.engine.reset()
  }

  public createMigration(
    params: EngineArgs.CreateMigrationInput,
  ): Promise<EngineResults.CreateMigrationOutput> {
    return this.engine.createMigration(params)
  }

  public diagnoseMigrationHistory({
    optInToShadowDatabase,
  }: {
    optInToShadowDatabase: boolean
  }): Promise<EngineResults.DiagnoseMigrationHistoryOutput> {
    return this.engine.diagnoseMigrationHistory({
      migrationsDirectoryPath: this.migrationsDirectoryPath,
      optInToShadowDatabase,
    })
  }

  public async getDbInfo(): Promise<{
    schemaWord: string
    dbType: string
    dbName: string
    dbLocation: string
  }> {
    const datamodel = this.getDatamodel()
    const config = await getConfig({ datamodel })
    const activeDatasource = config.datasources[0]
    const credentials = uriToCredentials(activeDatasource.url.value)
    const dbLocation = getDbLocation(credentials)
    return {
      ...getDbinfoFromCredentials(credentials),
      dbLocation,
    }
  }

  public listMigrationDirectories(): Promise<EngineResults.ListMigrationDirectoriesOutput> {
    return this.engine.listMigrationDirectories({
      migrationsDirectoryPath: this.migrationsDirectoryPath,
    })
  }

  public async markMigrationApplied({
    migrationId,
  }: {
    migrationId: string
  }): Promise<void> {
    return await this.engine.markMigrationApplied({
      migrationsDirectoryPath: this.migrationsDirectoryPath,
      migrationName: migrationId,
    })
  }

  public markMigrationRolledBack({
    migrationId,
  }: {
    migrationId: string
  }): Promise<void> {
    return this.engine.markMigrationRolledBack({
      migrationName: migrationId,
    })
  }

  public applyScript({ script }: { script: string }): Promise<void> {
    return this.engine.applyScript({ script })
  }

  public applyMigrations(): Promise<EngineResults.ApplyMigrationsOutput> {
    return this.engine.applyMigrations({
      migrationsDirectoryPath: this.migrationsDirectoryPath,
    })
  }

  public evaluateDataLoss(): Promise<EngineResults.EvaluateDataLossOutput> {
    const datamodel = this.getDatamodel()

    return this.engine.evaluateDataLoss({
      migrationsDirectoryPath: this.migrationsDirectoryPath,
      prismaSchema: datamodel,
    })
  }

  public async push({
    force = false,
  }: {
    force?: boolean
  }): Promise<EngineResults.SchemaPush> {
    const datamodel = this.getDatamodel()

    const {
      warnings,
      unexecutable,
      executedSteps,
    } = await this.engine.schemaPush({
      force,
      schema: datamodel,
    })

    return {
      executedSteps,
      warnings,
      unexecutable,
    }
  }

  public async tryToRunGenerate(): Promise<void> {
    const message: string[] = []

    console.info() // empty line
    logUpdate(
      `Running generate... ${chalk.dim(
        '(Use --skip-generate to skip the generators)',
      )}`,
    )

    const generators = await getGenerators({
      schemaPath: this.schemaPath,
      printDownloadProgress: false,
      version: enginesVersion,
      cliVersion: packageJson.version,
    })

    for (const generator of generators) {
      const toStr = generator.options!.generator.output!
        ? chalk.dim(
            ` to .${path.sep}${path.relative(
              process.cwd(),
              generator.options!.generator.output,
            )}`,
          )
        : ''
      const name = generator.manifest
        ? generator.manifest.prettyName
        : generator.options!.generator.provider

      logUpdate(`Running generate... - ${name}`)

      const before = Date.now()
      try {
        await generator.generate()
        const after = Date.now()
        const version = generator.manifest?.version
        message.push(
          `âœ” Generated ${chalk.bold(name!)}${
            version ? ` (${version})` : ''
          }${toStr} in ${formatms(after - before)}`,
        )
        generator.stop()
      } catch (err) {
        message.push(`${err.message}`)
        generator.stop()
      }
    }

    logUpdate(message.join('\n'))
  }
}
