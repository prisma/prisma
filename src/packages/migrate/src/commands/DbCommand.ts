import {
  arg,
  Command,
  Commands,
  format,
  HelpError,
  isError,
  unknownCommand,
  link,
} from '@prisma/sdk'
import chalk from 'chalk'

export class DbCommand implements Command {
  public static new(cmds: Commands): DbCommand {
    return new DbCommand(cmds)
  }

  private static help = format(`
${
  process.platform === 'win32' ? '' : chalk.bold('üèãÔ∏è  ')
}Manage your database schema and lifecycle during development.

${chalk.bold.yellow('WARNING')} ${chalk.bold(
    `Prisma db is currently in Preview (${link('https://pris.ly/d/preview')}).
There may be bugs and it's not recommended to use it in production environments.`,
  )}
${chalk.dim(
  'When using any of the subcommands below you need to explicitly opt-in via the --preview-feature flag.',
)}

${chalk.bold('Flag')}

  --preview-feature   Run preview Prisma commands

${chalk.bold('Usage')}

  With an existing schema.prisma:
  ${chalk.dim('$')} prisma db [command] [options] --preview-feature

  Or specify a schema path:
  ${chalk.dim(
    '$',
  )} prisma db [command] [options] --preview-feature --schema=./schema.prisma

${chalk.bold('Options')}

  -h, --help   Display this help message

${chalk.bold('Commands')}

        push   Push the state from Prisma schema to the database during prototyping ${chalk.dim(
          '(preview)',
        )}

${chalk.bold('Examples')}

  Using prisma db push
  ${chalk.dim('$')} prisma db push --preview-feature
  `)
  private constructor(private readonly cmds: Commands) {}

  public async parse(argv: string[]): Promise<string | Error> {
    const args = arg(argv, {
      '--help': Boolean,
      '-h': '--help',
      '--preview-feature': Boolean,
      '--telemetry-information': String,
    })

    if (isError(args)) {
      return this.help(args.message)
    }

    // display help for help flag or no subcommand
    if (args._.length === 0 || args['--help']) {
      return this.help()
    }

    // check if we have that subcommand
    const cmd = this.cmds[args._[0]]
    if (cmd) {
      const argsForCmd = args['--preview-feature']
        ? [...args._.slice(1), `--preview-feature`]
        : args._.slice(1)
      return cmd.parse(argsForCmd)
    }

    return unknownCommand(DbCommand.help, args._[0])
  }

  public help(error?: string): string | HelpError {
    if (error) {
      return new HelpError(
        `\n${chalk.bold.red(`!`)} ${error}\n${DbCommand.help}`,
      )
    }
    return DbCommand.help
  }
}
