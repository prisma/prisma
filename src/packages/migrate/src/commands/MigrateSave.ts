import {
  arg,
  Command,
  format,
  getCommandWithExecutor,
  getSchema,
  getSchemaDir,
  getSchemaPath,
  HelpError,
  isError,
} from '@prisma/sdk'
import chalk from 'chalk'
import fs from 'fs'
import path from 'path'
import prompt from 'prompts'
import { promisify } from 'util'
import { Migrate } from '../Migrate'
import { ensureDatabaseExists } from '../utils/ensureDatabaseExists'
import { ExperimentalFlagError } from '../utils/experimental'
import { printFiles } from '../utils/printFiles'
import { printMigrationId } from '../utils/printMigrationId'
import { serializeFileMap } from '../utils/serializeFileMap'

const writeFile = promisify(fs.writeFile)

/**
 * $ prisma migrate save
 */
export class MigrateSave implements Command {
  public static new(): MigrateSave {
    return new MigrateSave()
  }

  // static help template
  private static help = format(`
    Save a migration

    ${chalk.bold.yellow('WARNING')} ${chalk.bold(
    "Prisma's migration functionality is currently in an experimental state.",
  )}
    ${chalk.dim(
      'When using any of the commands below you need to explicitly opt-in via the --experimental flag.',
    )}

    ${chalk.bold('Usage')}

      ${chalk.dim('$')} prisma migrate save [options] --experimental

    ${chalk.bold('Options')}

      -h, --help       Displays this help message
      -n, --name       Name the migration
      -c, --create-db  Create the database in case it doesn't exist
      -p, --preview    Get a preview of which migration would be created next

    ${chalk.bold('Examples')}

      Create a new migration
      ${chalk.dim('$')} prisma migrate save --experimental

      Create a new migration by name
      ${chalk.dim(
        '$',
      )} prisma migrate save --name "add unique to email" --experimental

  `)

  // parse arguments
  public async parse(argv: string[]): Promise<string | Error> {
    // parse the arguments according to the spec
    const args = arg(argv, {
      '--help': Boolean,
      '-h': '--help',
      '--name': String,
      '-n': '--name',
      '--preview': Boolean,
      '-p': '--preview',
      '--create-db': Boolean,
      '-c': '--create-db',
      '--experimental': Boolean,
      '--schema': String,
      '--telemetry-information': String,
    })

    if (isError(args)) {
      return this.help(args.message)
    }

    if (args['--help']) {
      return this.help()
    }

    if (!args['--experimental']) {
      throw new ExperimentalFlagError()
    }

    const preview = args['--preview'] || false
    const schemaPath = await getSchemaPath(args['--schema'])

    if (!schemaPath) {
      throw new Error(
        `Could not find a ${chalk.bold(
          'schema.prisma',
        )} file that is required for this command.\nYou can either provide it with ${chalk.greenBright(
          '--schema',
        )}, set it as \`prisma.schema\` in your package.json or put it into the default location ${chalk.greenBright(
          './prisma/schema.prisma',
        )} https://pris.ly/d/prisma-schema-location`,
      )
    }

    console.log(
      chalk.dim(
        `Prisma schema loaded from ${path.relative(process.cwd(), schemaPath)}`,
      ),
    )

    await ensureDatabaseExists('create', args['--create-db'], schemaPath)

    const migrate = new Migrate(schemaPath)

    const migration = await migrate.createMigration('DUMMY_NAME')

    if (!migration) {
      migrate.stop()
      return `Everything up-to-date\n` // TODO: find better wording
    }

    const name = preview ? args['--name'] : await this.name(args['--name'])

    const { files, newLockFile, migrationId } = await migrate.save(
      migration,
      name,
      preview,
    )
    migrate.stop()

    if (
      migration.unexecutableMigrations &&
      migration.unexecutableMigrations.length > 0
    ) {
      const messages: string[] = []
      messages.push(
        `${chalk.bold(
          '\n\n⚠️ We found migration(s) that cannot be executed:\n',
        )}`,
      )
      for (const item of migration.unexecutableMigrations) {
        messages.push(`${chalk(`  • ${item.description}`)}`)
      }
      console.log() // empty line
      throw new Error(`${messages.join('\n')}\n`)
    }

    if (migration.warnings && migration.warnings.length > 0) {
      console.log(
        chalk.bold(
          `\n\n⚠️  There might be data loss when applying the migration:\n`,
        ),
      )
      for (const warning of migration.warnings) {
        console.log(chalk(`  • ${warning.description}`))
      }
      console.log() // empty line
    }

    if (preview) {
      return `\nRun ${chalk.greenBright(
        getCommandWithExecutor(
          'prisma migrate save --name MIGRATION_NAME --experimental',
        ),
      )} to create the migration\n`
    }

    await getSchema(schemaPath) // just to leverage on its error handling
    const schemaDir = (await getSchemaDir(schemaPath))! // TODO: Probably getSchemaDir() should return Promise<string> instead of Promise<string | null>

    const migrationsDir = path.join(schemaDir, 'migrations', migrationId)
    await serializeFileMap(files, migrationsDir)
    const lockFilePath = path.join(schemaDir, 'migrations', 'migrate.lock')
    await writeFile(lockFilePath, newLockFile)

    return `\nPrisma Migrate just created your migration ${printMigrationId(
      migrationId,
    )} in\n\n${chalk.dim(
      printFiles(`migrations/${migrationId}`, files),
    )}\n\nRun ${chalk.greenBright(
      getCommandWithExecutor('prisma migrate up --experimental'),
    )} to apply the migration\n`
  }

  // get the name
  public async name(name?: string): Promise<string | undefined> {
    if (name === '') {
      return undefined
    }
    if (name) {
      return name
    }
    const response = await prompt({
      type: 'text',
      name: 'name',
      message: `Name of migration`,
    })
    return response.name || undefined
  }

  // help message
  public help(error?: string): string | HelpError {
    if (error) {
      return new HelpError(
        `\n${chalk.bold.red(`!`)} ${error}\n${MigrateSave.help}`,
      )
    }
    return MigrateSave.help
  }
}
