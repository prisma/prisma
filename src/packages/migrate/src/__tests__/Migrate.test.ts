import assert from 'assert'
import del from 'del'
import mkdir from 'make-dir'
import fs from 'fs'
import { promisify } from 'util'
import { dirname, join } from 'path'
import tempy from 'tempy'
import dedent from 'strip-indent'
import Database from 'sqlite-async'
import stripAnsi from 'strip-ansi'
import { Migrate } from '../Migrate'
import { MigrateSave } from '../commands/MigrateSave'

const writeFile = promisify(fs.writeFile)
const testRootDir = tempy.directory()

describe('migrate.create', () => {
  beforeEach(async () => {
    await mkdir(testRootDir)
  })

  afterEach(async () => {
    await del(testRootDir, { force: true }) // Need force: true because `del` does not delete dirs outside the CWD
  })

  createTests().map((t) => {
    // eslint-disable-next-line jest/expect-expect
    test(t.name, async () => {
      const schemaPath = join(testRootDir, Object.keys(t.fs)[0])
      await writeFiles(testRootDir, t.fs)
      await t.fn(schemaPath)
    })
  })
})

// create a temporary set of files
async function writeFiles(
  root: string,
  files: {
    [name: string]: any // eslint-disable-line @typescript-eslint/no-explicit-any
  },
): Promise<string> {
  for (const name in files) {
    const filepath = join(root, name)
    await mkdir(dirname(filepath))
    await writeFile(filepath, dedent(files[name]))
  }
  // return the test path
  return root
}

function replaceTimestamp(string: string): string {
  const regex = /[0-9]{14}/gm
  return string.replace(regex, 'X'.repeat(14))
}

// create file tests
// eslint-disable-next-line @typescript-eslint/explicit-function-return-type
function createTests() {
  return [
    {
      name: 'simple ok',
      fs: {
        'schema.prisma': `
          datasource my_db {
            provider = "sqlite"
            url = "file:./db/db_file.db"
            default = true
          }

          model User {
            id Int @id
          }
        `,
        'db/.keep': ``,
      },
      fn: async (schemaPath: string): Promise<undefined> => {
        const migrate = new Migrate(schemaPath)
        const migration = await migrate.createMigration('setup')
        const result = await migrate.save(migration!, 'setup')
        migrate.stop()
        if (typeof result === 'undefined') {
          return assert.fail(`result shouldn't be undefined`)
        }
        assert.ok(result.migrationId.includes('-setup'))
        assert.ok(result.newLockFile)
        assert.ok(result.files['steps.json'])
        assert.ok(result.files['schema.prisma'])
        assert.ok(result.files['README.md'])
        expect(migration?.datamodelSteps).toMatchSnapshot()
        expect(migration?.warnings).toMatchSnapshot()
        expect(migration?.unexecutableMigrations).toMatchSnapshot()
      },
    },
    {
      name: 'spaces ok',
      fs: {
        'schema.prisma': `
          datasource my_db {
            provider = "sqlite"
            url = "file:./db/db_file.db"
            default = true
          }

          model User {
            id Int @id
          }
        `,
        'db/.keep': ``,
      },
      fn: async (schemaPath: string): Promise<undefined> => {
        const migrate = new Migrate(schemaPath)
        const migration = await migrate.createMigration('initial setup')
        const result = await migrate.save(migration!, 'initial setup')
        migrate.stop()
        if (typeof result === 'undefined') {
          return assert.fail(`result shouldn't be undefined`)
        }
        assert.ok(result.migrationId.includes(`-initial-setup`))
        assert.ok(result.newLockFile)
        assert.ok(result.files['steps.json'])
        assert.ok(result.files['schema.prisma'])
        assert.ok(result.files['README.md'])
        expect(migration?.datamodelSteps).toMatchSnapshot()
        expect(migration?.warnings).toMatchSnapshot()
        expect(migration?.unexecutableMigrations).toMatchSnapshot()
      },
    },
    {
      name: 'dashes ok',
      fs: {
        'schema.prisma': `
          datasource my_db {
            provider = "sqlite"
            url = "file:./db/db_file.db"
            default = true
          }

          model User {
            id Int @id
          }
        `,
        'db/.keep': ``,
      },
      fn: async (schemaPath: string): Promise<undefined> => {
        const migrate = new Migrate(schemaPath)
        const migration = await migrate.createMigration('initial setup')
        const result = await migrate.save(migration!, 'initial setup')
        migrate.stop()
        if (typeof result === 'undefined') {
          return assert.fail(`result shouldn't be undefined`)
        }
        assert.ok(result.migrationId.includes(`-initial-setup`))
        assert.ok(result.newLockFile)
        assert.ok(result.files['steps.json'])
        assert.ok(result.files['schema.prisma'])
        assert.ok(result.files['README.md'])
        expect(migration?.datamodelSteps).toMatchSnapshot()
        expect(migration?.warnings).toMatchSnapshot()
        expect(migration?.unexecutableMigrations).toMatchSnapshot()
      },
    },
    {
      name: 'custom schema filename ok',
      fs: {
        'myawesomeschema.file': `
          datasource my_db {
            provider = "sqlite"
            url = "file:./db/db_file.db"
            default = true
          }

          model User {
            id Int @id
          }
        `,
        'db/.keep': ``,
      },
      fn: async (schemaPath: string): Promise<undefined> => {
        const migrate = new Migrate(schemaPath)
        const migration = await migrate.createMigration('setup')
        const result = await migrate.save(migration!, 'setup')
        migrate.stop()
        if (typeof result === 'undefined') {
          return assert.fail(`result shouldn't be undefined`)
        }
        assert.ok(result.migrationId.includes('-setup'))
        assert.ok(result.newLockFile)
        assert.ok(result.files['steps.json'])
        assert.ok(result.files['schema.prisma'])
        assert.ok(result.files['README.md'])
        expect(migration?.datamodelSteps).toMatchSnapshot()
        expect(migration?.warnings).toMatchSnapshot()
        expect(migration?.unexecutableMigrations).toMatchSnapshot()
      },
    },
    {
      name: 'custom folder and schema filename name ok',
      fs: {
        'awesome/myawesomeschema.file': `
          datasource my_db {
            provider = "sqlite"
            url = "file:../db/db_file.db"
            default = true
          }

          model User {
            id Int @id
          }
        `,
        'db/.keep': ``,
      },
      fn: async (schemaPath: string): Promise<undefined> => {
        const migrate = new Migrate(schemaPath)
        const migration = await migrate.createMigration('setup')
        const result = await migrate.save(migration!, 'setup')
        migrate.stop()
        if (typeof result === 'undefined') {
          return assert.fail(`result shouldn't be undefined`)
        }
        assert.ok(result.migrationId.includes('-setup'))
        assert.ok(result.newLockFile)
        assert.ok(result.files['steps.json'])
        assert.ok(result.files['schema.prisma'])
        assert.ok(result.files['README.md'])
        expect(migration?.datamodelSteps).toMatchSnapshot()
        expect(migration?.warnings).toMatchSnapshot()
        expect(migration?.unexecutableMigrations).toMatchSnapshot()
      },
    },
    {
      name: 'invalid ok',
      fs: {
        'schema.prisma': `
          datasource my_db {
            provider = "sqlite"
            url = "file:./db/db_file.db"
            default = true
          }

          model User {
            id Int @id
            canBeNull String?
          }
        `,
        'schema-not-null.prisma': `
          datasource my_db {
            provider = "sqlite"
            url = "file:./db/db_file.db"
            default = true
          }

          model User {
            id Int @id
            canBeNull String
            requiredSomething String
          }
        `,
        'db/.keep': ``,
      },
      fn: async (schemaPath: string): Promise<undefined> => {
        const migrate = new Migrate(schemaPath)
        const migration = await migrate.createMigration('setup1')
        const result = await migrate.save(migration!, 'setup1')
        if (typeof result === 'undefined') {
          return assert.fail(`result shouldn't be undefined`)
        }
        assert.ok(result.migrationId.includes('-setup1'))
        assert.ok(result.newLockFile)
        assert.ok(result.files['steps.json'])
        assert.ok(result.files['schema.prisma'])
        assert.ok(result.files['README.md'])
        expect(migration?.datamodelSteps).toMatchSnapshot()
        expect(migration?.warnings).toMatchSnapshot()
        expect(migration?.unexecutableMigrations).toMatchSnapshot()

        // Save from CLI - write files to filesystem
        const resultSave = await MigrateSave.new().parse([
          `--schema=${schemaPath}`,
          `--name=init`,
          '--experimental',
        ])
        expect(
          replaceTimestamp(stripAnsi(resultSave as string)),
        ).toMatchSnapshot()

        await migrate.up()
        migrate.stop()

        /* eslint-disable @typescript-eslint/no-unsafe-assignment */
        /* eslint-disable @typescript-eslint/no-unsafe-call */
        /* eslint-disable @typescript-eslint/no-unsafe-member-access */
        const db = await Database.open(
          schemaPath.replace('schema.prisma', 'db/db_file.db'),
        )
        await db.exec('INSERT INTO User (canBeNull) VALUES ("Something!")')
        await db.close()
        /* eslint-enable @typescript-eslint/no-unsafe-assignment */
        /* eslint-enable @typescript-eslint/no-unsafe-call */
        /* eslint-enable @typescript-eslint/no-unsafe-member-access */

        const schemaPath2 = schemaPath.replace(
          'schema.prisma',
          'schema-not-null.prisma',
        )
        const migrate2 = new Migrate(schemaPath2)
        const migration2 = await migrate2.createMigration('setup2')
        const result2 = await migrate2.save(migration2!, 'setup2')
        migrate2.stop()
        if (typeof result2 === 'undefined') {
          return assert.fail(`result2 shouldn't be undefined`)
        }
        assert.ok(result2.migrationId.includes('-setup2'))
        assert.ok(result2.newLockFile)
        assert.ok(result2.files['steps.json'])
        assert.ok(result2.files['schema.prisma'])
        assert.ok(result2.files['README.md'])
        expect(migration2?.datamodelSteps).toMatchSnapshot()
        expect(migration2?.warnings).toMatchSnapshot()

        const oldConsoleLog = console.log
        const logs: string[] = []
        console.log = (...args) => {
          logs.push(...args)
        }

        try {
          // Save from CLI - write files to filesystem
          await MigrateSave.new().parse([
            `--schema=${schemaPath2}`,
            `--name=init-2`,
            '--experimental',
          ])
        } catch (e) {
          // Should error with unexecutableMigrations:
          /* eslint-disable-next-line @typescript-eslint/no-unsafe-member-access */
          expect(stripAnsi(e.message)).toMatchSnapshot()
        }
        console.log = oldConsoleLog
        expect(stripAnsi(logs.join('\n'))).toMatchSnapshot()
      },
    },
    {
      name: 'simple debug panic',
      fs: {
        'schema.prisma': `
      datasource my_db {
        provider = "sqlite"
        url = "file:./db/db_file.db"
        default = true
      }

      model User {
        id Int @id
      }
    `,
        'db/.keep': ``,
      },
      fn: async (schemaPath: string): Promise<undefined> => {
        const migrate = new Migrate(schemaPath)

        try {
          await migrate.engine.debugPanic()
        } catch (e) {
          expect(
            // remove hash
            // eslint-disable-next-line @typescript-eslint/no-unsafe-member-access
            stripAnsi(e.message).replace(/\/rustc\/(.+)\//, '/rustc/hash/'),
          ).toMatchSnapshot()
        } finally {
          migrate.stop()
        }
        return
      },
    },
  ]
}
