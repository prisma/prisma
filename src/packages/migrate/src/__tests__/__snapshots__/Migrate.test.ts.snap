// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`migrate.create custom folder and schema filename name ok 1`] = `
Array [
  Object {
    source: my_db,
    tag: CreateSource,
  },
  Object {
    argument: provider,
    location: Object {
      source: my_db,
      tag: Source,
    },
    tag: CreateArgument,
    value: "sqlite",
  },
  Object {
    argument: url,
    location: Object {
      source: my_db,
      tag: Source,
    },
    tag: CreateArgument,
    value: "***",
  },
  Object {
    argument: default,
    location: Object {
      source: my_db,
      tag: Source,
    },
    tag: CreateArgument,
    value: true,
  },
  Object {
    model: User,
    tag: CreateModel,
  },
  Object {
    arity: Required,
    field: id,
    model: User,
    tag: CreateField,
    type: Int,
  },
  Object {
    location: Object {
      directive: id,
      path: Object {
        field: id,
        model: User,
        tag: Field,
      },
    },
    tag: CreateDirective,
  },
]
`;

exports[`migrate.create custom folder and schema filename name ok 2`] = `Array []`;

exports[`migrate.create custom folder and schema filename name ok 3`] = `Array []`;

exports[`migrate.create custom schema filename ok 1`] = `
Array [
  Object {
    source: my_db,
    tag: CreateSource,
  },
  Object {
    argument: provider,
    location: Object {
      source: my_db,
      tag: Source,
    },
    tag: CreateArgument,
    value: "sqlite",
  },
  Object {
    argument: url,
    location: Object {
      source: my_db,
      tag: Source,
    },
    tag: CreateArgument,
    value: "***",
  },
  Object {
    argument: default,
    location: Object {
      source: my_db,
      tag: Source,
    },
    tag: CreateArgument,
    value: true,
  },
  Object {
    model: User,
    tag: CreateModel,
  },
  Object {
    arity: Required,
    field: id,
    model: User,
    tag: CreateField,
    type: Int,
  },
  Object {
    location: Object {
      directive: id,
      path: Object {
        field: id,
        model: User,
        tag: Field,
      },
    },
    tag: CreateDirective,
  },
]
`;

exports[`migrate.create custom schema filename ok 2`] = `Array []`;

exports[`migrate.create custom schema filename ok 3`] = `Array []`;

exports[`migrate.create dashes ok 1`] = `
Array [
  Object {
    source: my_db,
    tag: CreateSource,
  },
  Object {
    argument: provider,
    location: Object {
      source: my_db,
      tag: Source,
    },
    tag: CreateArgument,
    value: "sqlite",
  },
  Object {
    argument: url,
    location: Object {
      source: my_db,
      tag: Source,
    },
    tag: CreateArgument,
    value: "***",
  },
  Object {
    argument: default,
    location: Object {
      source: my_db,
      tag: Source,
    },
    tag: CreateArgument,
    value: true,
  },
  Object {
    model: User,
    tag: CreateModel,
  },
  Object {
    arity: Required,
    field: id,
    model: User,
    tag: CreateField,
    type: Int,
  },
  Object {
    location: Object {
      directive: id,
      path: Object {
        field: id,
        model: User,
        tag: Field,
      },
    },
    tag: CreateDirective,
  },
]
`;

exports[`migrate.create dashes ok 2`] = `Array []`;

exports[`migrate.create dashes ok 3`] = `Array []`;

exports[`migrate.create invalid ok 1`] = `
Array [
  Object {
    source: my_db,
    tag: CreateSource,
  },
  Object {
    argument: provider,
    location: Object {
      source: my_db,
      tag: Source,
    },
    tag: CreateArgument,
    value: "sqlite",
  },
  Object {
    argument: url,
    location: Object {
      source: my_db,
      tag: Source,
    },
    tag: CreateArgument,
    value: "***",
  },
  Object {
    argument: default,
    location: Object {
      source: my_db,
      tag: Source,
    },
    tag: CreateArgument,
    value: true,
  },
  Object {
    model: User,
    tag: CreateModel,
  },
  Object {
    arity: Required,
    field: id,
    model: User,
    tag: CreateField,
    type: Int,
  },
  Object {
    location: Object {
      directive: id,
      path: Object {
        field: id,
        model: User,
        tag: Field,
      },
    },
    tag: CreateDirective,
  },
  Object {
    arity: Optional,
    field: canBeNull,
    model: User,
    tag: CreateField,
    type: String,
  },
]
`;

exports[`migrate.create invalid ok 2`] = `Array []`;

exports[`migrate.create invalid ok 3`] = `Array []`;

exports[`migrate.create invalid ok 4`] = `

Prisma Migrate just created your migration XXXXXXXXXXXXXX-init in

migrations/
  ‚îî‚îÄ XXXXXXXXXXXXXX-init/
    ‚îî‚îÄ steps.json
    ‚îî‚îÄ schema.prisma
    ‚îî‚îÄ README.md

Run prisma migrate up --experimental to apply the migration

`;

exports[`migrate.create invalid ok 5`] = `
Array [
  Object {
    arity: Required,
    field: requiredSomething,
    model: User,
    tag: CreateField,
    type: String,
  },
  Object {
    arity: Required,
    field: canBeNull,
    model: User,
    tag: UpdateField,
  },
]
`;

exports[`migrate.create invalid ok 6`] = `Array []`;

exports[`migrate.create invalid ok 7`] = `


‚ö†Ô∏è We found migration(s) that cannot be executed:

  ‚Ä¢ Added the required column \`requiredSomething\` to the \`User\` table without a default value. There are 1 rows in this table, it is not possible to execute this migration.

`;

exports[`migrate.create invalid ok 8`] = `
Prisma schema loaded from schema-not-null.prisma
üìº  migrate save --name init-2

Local datamodel Changes:

model User {
  id Int @id
  canBeNull String?
  canBeNull String
  requiredSomething String
}
`;

exports[`migrate.create simple debug panic 1`] = `
Error in migration engine.
Reason: [/rustc/hash/macros.rs:0:0] This is the debugPanic artificial panic

Please create an issue in the migrate repo with
your \`schema.prisma\` and the prisma command you tried to use üôè:
https://github.com/prisma/migrate/issues/new

`;

exports[`migrate.create simple ok 1`] = `
Array [
  Object {
    source: my_db,
    tag: CreateSource,
  },
  Object {
    argument: provider,
    location: Object {
      source: my_db,
      tag: Source,
    },
    tag: CreateArgument,
    value: "sqlite",
  },
  Object {
    argument: url,
    location: Object {
      source: my_db,
      tag: Source,
    },
    tag: CreateArgument,
    value: "***",
  },
  Object {
    argument: default,
    location: Object {
      source: my_db,
      tag: Source,
    },
    tag: CreateArgument,
    value: true,
  },
  Object {
    model: User,
    tag: CreateModel,
  },
  Object {
    arity: Required,
    field: id,
    model: User,
    tag: CreateField,
    type: Int,
  },
  Object {
    location: Object {
      directive: id,
      path: Object {
        field: id,
        model: User,
        tag: Field,
      },
    },
    tag: CreateDirective,
  },
]
`;

exports[`migrate.create simple ok 2`] = `Array []`;

exports[`migrate.create simple ok 3`] = `Array []`;

exports[`migrate.create spaces ok 1`] = `
Array [
  Object {
    source: my_db,
    tag: CreateSource,
  },
  Object {
    argument: provider,
    location: Object {
      source: my_db,
      tag: Source,
    },
    tag: CreateArgument,
    value: "sqlite",
  },
  Object {
    argument: url,
    location: Object {
      source: my_db,
      tag: Source,
    },
    tag: CreateArgument,
    value: "***",
  },
  Object {
    argument: default,
    location: Object {
      source: my_db,
      tag: Source,
    },
    tag: CreateArgument,
    value: true,
  },
  Object {
    model: User,
    tag: CreateModel,
  },
  Object {
    arity: Required,
    field: id,
    model: User,
    tag: CreateField,
    type: Int,
  },
  Object {
    location: Object {
      directive: id,
      path: Object {
        field: id,
        model: User,
        tag: Field,
      },
    },
    tag: CreateDirective,
  },
]
`;

exports[`migrate.create spaces ok 2`] = `Array []`;

exports[`migrate.create spaces ok 3`] = `Array []`;
